qmail-alias.8
.TH qmail-alias 8
.SH NAME
qmail-alias \- deliver or forward a mail message
.SH SYNOPSIS
.B qmail-alias
[
.B \-nN
]
.I user
.I homedir
.I local
.I dash
.I ext
.I domain
.I sender
.SH DESCRIPTION
.B qmail-alias
reads a mail message
and delivers it to
.I user
by the procedure described in
.BR dot-qmail(5) .

The message's envelope recipient is
.IR local@domain .
.B qmail-alias
records
.I local@domain
in a new
.B Delivered-To
header field.
If exactly the same
.B Delivered-To: \fIlocal@domain
already appears in the header,
.B qmail-alias
bounces the message,
to prevent mail forwarding loops.

The message's envelope sender is
.IR sender .
.B qmail-alias
records
.I sender
in a new
.B Return-Path
header field.

.I homedir
is the user's home directory.
It must be an absolute directory name.

.I dash
and
.I ext
identify the
.B .qmail\fIdashext
file used by
.BR qmail-alias ;
see
.BR dot-qmail(5) .
Normally
.I dash
is either empty or a lone hyphen.
If it is empty,
.B qmail-alias
treats a nonexistent
.B .qmail\fIext
the same way as an empty
.BR .qmail\fIext .

The standard input for
.B qmail-alias
must be a seekable file,
so that
.B qmail-alias
can read it more than once.
.SH "OPTIONS"
.TP
.B \-n
Instead of reading and delivering the message,
print a description of the delivery instructions.
.TP
.B \-N
(Default.) Read and deliver the message.
.SH "EXIT CODES"
0 if the delivery is completely successful;
nonzero if any delivery instruction failed.
Exit code 111
indicates temporary failure.
.SH "SEE ALSO"
dot-qmail(5),
envelopes(5),
qmail-command(8),
qmail-queue(8),
qmail-send(8),
qmail-lspawn(8)


CHANGES
19970220 qmail 1.00.
19970219 change: various documentation tweaks.
19970218 change: updated THOUGHTS.
19970218 change: talked about SPAWN_NUMD in FAQ. tnx EC.
19970210 change: noted in maildir.5 that readers should skip any name
         starting with a dot. tnx SP.
19970209 change: added note to splogger.8 about reliability. tnx BT.
19970209 change: added section to FAQ on slow sendmail switch. tnx BT.
19970206 change: added section to FAQ about dtcm. tnx PJG.
19970206 change: tweaked maildir.5.
19970201 change: added MH spost note to FAQ. tnx TU.
19970131 change: reorganized FAQ.
19970131 change: added web references to FAQ.
19970124 change: tweaked qmail-upgrade man page.
19970120 qmail 0.96, gamma.
19970120 change: removed various try* in auto-configuration.
19970120 bug: qmail-inject fails to quote argument addresses. impact:
         addresses containing special characters could be misinterpreted
	 or rejected. tnx C2F. fix: use quote2().
19970120 portability problem: ESIX puts syslog() and openlog() into
         -lgen. impact: can't compile under ESIX. fix: put -lgen into
	 LIBS for unix_sv. tnx RN.
19961221 qmail 0.95, gamma.
19961218 change: added various try* to TARGETS. tnx SA.
19961216 change: clarified in qmail-send.8 that virtualdomains does not
         apply to domains listed in locals.
19961216 change: slurpclose() now closes fd on out-of-memory. makes it
         more widely applicable.
19961215 change: replaced elm instructions in INSTALL.mbox with an
         explanation of what source change to make. tnx AB.
19961212 portability problem: under NEWS-OS, time_t needs sys/types.h.
         impact: couldn't compile under NEWS-OS. fix: include
	 sys/types.h in predate.c. tnx TU.
19961211 change: used timeoutread, timeoutwrite in remoteinfo(). tnx
         REB.
19961210 portability problem: apparently some SGIs produce a systype of
         irix64. impact: couldn't compile on those systems. fix: handle
	 irix64 in make-cmds. tnx M3S.
19961208 change: added note to maildir2mbox.1 about mbox locking.
19961208 qmail 0.94, gamma.
19961207 change: added QMAILDEFAULTDOMAIN, QMAILDEFAULTHOST,
         QMAILIDHOST, QMAILPLUSDOMAIN. tnx BTW.
19961206 cleanup: readsubdir() protects itself against name overflow,
         rather than depending on caller.
19961204 change: changed FAQ 7.3 to prohibit fixup relaying.
19961203 change: added note to FAQ about possibly having to put a space
         before "$SENDER" for uux. tnx FW.
19961202 change: added FAQ entry on QUEUE_EXTRA.
19961202 change: added FAQ entry on backups. tnx DP.
19961202 change: added note to INSTALL.mbox about qpopper. tnx VV.
19961201 change: replaced logger with splogger in FAQ 5.1. tnx FPL.
19961201 change: used netmask example for tcpcontrol in FAQ. tnx FPL.
19961201 change: added note to README about the mailing list. tnx FPL.
19961201 change: documented rcpthosts wildcards. tnx RN.
19961201 change: added note to FAQ about making mailx use datemail.
19961201 change: added datemail. function requested by several people;
         approach suggested by TG.
19961201 change: added predate.
19961129 change: added QUEUE_EXTRA, QUEUE_EXTRALEN.
19961129 change: qmail-remote bounces messages with partial final lines.
19961129 change: added atomcheck() to quote crappy atoms.
19961129 change: revised atomok() to let atoms deal with more crap.
19961127 change: qmail-send adds note to deferral if flagdying. tnx TG.
19961127 change: split off maildirbounce, maildir2qmtp, and maildir2smtp
         into a separate serialmail package.
19961126 change: eliminated beta success reports from README.
19961124 change: forced res_query() return value to fit inside incoming
         buffer size. allegedly there are buggy versions of res_query()
	 that don't guarantee this.
19961122 qmail 0.93, gamma.
19961122 change: allowed empty arg list in forward.
19961121 change: qmail-smtpd now uses unknown (like qmail-qmtpd) rather
         than dying if environment variables are not set.
19961121 cleanup: reorganized helo handling in qmail-smtpd.c.
19961121 cleanup: eliminated newfield_rec.
19961121 cleanup: introduced DATE822FMT. used it in received.c.
19961121 cleanup: introduced received.c. used it in qmail-qmtpd.c,
         qmail-smtpd.c.
19961121 change: qmail-smtpd now generates a new timestamp for each
         message. tnx PJG.
19961121 cleanup: used stralloc in newfield.
19961121 cleanup: eliminated newfield_cc.
19961121 change: eliminated hfield_mort().
19961119 change: added 2-minute damper on tcpto.
19961118 change: wrote defaults for readfile controls in showctl.
19961117 change: control_readfile() now allows comments and blank lines.
         tnx LW.
19961117 change: qmail-start sets logger gid to GID_NOFILES.
19961117 bug: ipme_init() uses a fixed-length buffer for SIOCGIFCONF.
         impact: qmail-smtpd and qmail-remote will die if there are too
	 many local IP addresses. tnx MD. fix: ipme_init() now
	 dynamically allocates space, up to 200000 bytes, as long as
	 SIOCGIFCONF keeps failing. note that this is a very widespread
	 bug; it's in amd, exim, mrouted, named, nntpd, rarpd, sendmail,
	 tcpdump, timed, xntpd, and probably dozens more programs.
19961117 portability problem: on BSD 4.4 and various other systems,
         SIOCGIFCONF will truncate long lists and return success.
	 impact: on those systems, qmail-smtpd and qmail-remote will
	 miss some local IP addresses if there are too many. fix:
	 ipme_init() now checks whether there is enough space left in
	 the buffer for another ifreq, plus 64 bytes JIC. yuck.
19961117 change: ipmeprint now flushes only at end.
19961117 cleanup: introduced subfdinsmall. used it in qmail-clean.c,
         qmail-qmtpd.c.
19961117 cleanup: introduced subfdoutsmall. used it in hostname.c,
         printbreak.c, printnumd.c, printsplit.c, qmail-alias.c,
	 qmail-clean.c, qmail-getpw.c, qmail-qmtpd.c, maildir2smtp.c,
	 maildir2qmtp.c.
19961117 change: moved subfderr buf up to 256 characters.

qmail-alias.c
#include <sys/types.h>
#include <sys/stat.h>
#include "readwrite.h"
#include "signal.h"
#include "env.h"
#include "byte.h"
#include "exit.h"
#include "fork.h"
#include "open.h"
#include "wait.h"
#include "lock.h"
#include "seek.h"
#include "substdio.h"
#include "getline.h"
#include "subfd.h"
#include "sgetopt.h"
#include "alloc.h"
#include "error.h"
#include "stralloc.h"
#include "fmt.h"
#include "str.h"
#include "qqtalk.h"
#include "now.h"
#include "myctime.h"
#include "quote.h"
#include "conf-unusual.h"
#include "gfrom.h"
#include "case.h"
#include "slurpclose.h"

void err(s) char *s; { substdio_putsflush(subfderr,s); }
void soft() { _exit(111); }
void hard() { _exit(112); }

void temp_childcrashed() { err("Aack, child crashed. (#4.3.0)\n"); soft(); }
void temp_rewind() { err("Unable to rewind message. (#4.3.0)\n"); soft(); }
void temp_fork() { err("Unable to fork. (#4.3.0)\n"); soft(); }
void temp_read() { err("Error while reading message. (#4.3.0)\n"); soft(); }
void temp_write() { err("Error while writing message. (#4.3.0)\n"); soft(); }
void temp_child() { err("Temporary error in forwarding message. (#4.3.0)\n"); soft(); }
void temp_maildirtimeout() { err("Timeout on maildir delivery. (#4.3.0)\n"); soft(); }
void temp_maildir() { err("Temporary error on maildir delivery. (#4.3.0)\n"); soft(); }
void temp_nomaildir() { err("Unable to chdir to maildir. (#4.2.1)\n"); soft(); }
void temp_open(fn) char *fn; { err("Unable to open "); err(fn); err(". (#4.2.1)\n"); soft(); }

void temp_blankline() { err("Uh-oh: first line of .qmail file is blank. (#4.2.1)\n"); soft(); }
void temp_fofile() { err("Uh-oh: .qmail has file delivery but has x bit set. (#4.7.0)\n"); soft(); }
void temp_foprog() { err("Uh-oh: .qmail has prog delivery but has x bit set. (#4.7.0)\n"); soft(); }
void temp_nomem() { err("Out of memory. (#4.3.0)\n"); soft(); }
void temp_chdir() { err("Unable to switch to home directory. (#4.3.0)\n"); soft(); }
void temp_homestat() { err("Unable to stat home directory. (#4.3.0)\n"); soft(); }
void temp_homesticky() { err("Home directory is sticky: user is editing his .qmail file. (#4.2.1)\n"); soft(); }
void temp_homewritable() { err("Uh-oh: home directory is writable. (#4.7.0)\n"); soft(); }
void temp_qmwritable() { err("Uh-oh: .qmail file is writable. (#4.7.0)\n"); soft(); }
void temp_nfsqmail() { err("Temporary error trying to open .qmail file. (#4.3.0)\n"); soft(); }
void temp_denyqmail() { err("Permission error trying to open .qmail file. (#4.3.0)\n"); soft(); }
void temp_slowlock() { err("File has been locked for 30 seconds straight. (#4.3.0)\n"); soft(); }

void bounce_childperm() { err("Permanent error in forwarding message. (#5.2.4)\n"); hard(); }
void bounce_loop() { err("This message is looping: it already has my Delivered-To line. (#5.4.6)\n"); hard(); }
void bounce_ext() { err("Sorry, no mailbox here by that name. (#5.1.1)\n"); hard(); }
void usage() { err("qmail-alias: usage: qmail-alias [ -nN ] user homedir local dash ext domain sender\n"); hard(); }

void warn_homesticky() { err("Warning: home directory is sticky.\n"); }

int flagdoit;
int flag99;

char *user;
char *homedir;
char *local;
char *dash;
char *ext;
char *host;
char *sender;

stralloc dashext = {0};
stralloc ufline = {0};
stralloc rpline = {0};
stralloc envrecip = {0};
stralloc dtline = {0};
stralloc qme = {0};
stralloc ueo = {0};
stralloc cmds = {0};
stralloc messline = {0};
stralloc foo = {0};

char buf[1024];
char outbuf[1024];

/* child process */

char fntmptph[80 + FMT_ULONG * 2];
char fnnewtph[80 + FMT_ULONG * 2];
void tryunlinktmp() { unlink(fntmptph); }
void sigalrm() { tryunlinktmp(); _exit(3); }

void maildir_child(dir)
char *dir;
{
 unsigned long pid;
 unsigned long time;
 char host[64];
 char *s;
 int loop;
 struct stat st;
 int fd;
 substdio ss;
 substdio ssout;

 signal_catchalarm(sigalrm);
 if (chdir(dir) == -1) { if (error_temp(errno)) _exit(1); _exit(2); }
 pid = getpid();
 host[0] = 0;
 gethostname(host,sizeof(host));
 for (loop = 0;;++loop)
  {
   time = now();
   s = fntmptph;
   s += fmt_str(s,"tmp/");
   s += fmt_ulong(s,time); *s++ = '.';
   s += fmt_ulong(s,pid); *s++ = '.';
   s += fmt_strn(s,host,sizeof(host)); *s++ = 0;
   if (stat(fntmptph,&st) == -1) if (errno == error_noent) break;
   /* really should never get to this point */
   if (loop == 2) _exit(1);
   sleep(2);
  }
 str_copy(fnnewtph,fntmptph);
 byte_copy(fnnewtph,3,"new");

 alarm(86400);
 fd = open_excl(fntmptph);
 if (fd == -1) _exit(1);

 substdio_fdbuf(&ss,read,0,buf,sizeof(buf));
 substdio_fdbuf(&ssout,write,fd,outbuf,sizeof(outbuf));
 if (substdio_put(&ssout,rpline.s,rpline.len) == -1) goto fail;
 if (substdio_put(&ssout,dtline.s,dtline.len) == -1) goto fail;

 switch(substdio_copy(&ssout,&ss))
  {
   case -2: tryunlinktmp(); _exit(4);
   case -3: goto fail;
  }

 if (substdio_flush(&ssout) == -1) goto fail;
 if (fsync(fd) == -1) goto fail;
 if (close(fd) == -1) goto fail; /* NFS dorks */

 if (link(fntmptph,fnnewtph) == -1) goto fail;
   /* if it was error_exist, almost certainly successful; i hate NFS */
 tryunlinktmp(); _exit(0);

 fail: tryunlinktmp(); _exit(1);
}

/* end child process */

void maildir(fn)
char *fn;
{
 int child;
 int wstat;

 if (seek_begin(0) == -1) temp_rewind();

 switch(child = fork())
  {
   case -1:
     temp_fork();
   case 0:
     maildir_child(fn);
     soft();
  }

 wait_pid(&wstat,child);
 if (wait_crashed(wstat))
   temp_childcrashed();
 switch(wait_exitcode(wstat))
  {
   case 0: break;
   case 2: temp_nomaildir();
   case 3: temp_maildirtimeout();
   case 4: temp_read();
   default: temp_maildir();
  }
}

void slowlock() { temp_slowlock(); }

void mailfile(fn)
char *fn;
{
 int fd;
 substdio ss;
 substdio ssout;
 int match;
 seek_pos pos;
 int flaglocked;

 if (seek_begin(0) == -1) temp_rewind();

 fd = open_append(fn);
 if (fd == -1) temp_open(fn);

 signal_catchalarm(slowlock);
 alarm(30);
 flaglocked = (lock_ex(fd) != -1);
 alarm(0);
 signal_uncatchalarm();

 seek_end(fd);
 pos = seek_cur(fd);

 substdio_fdbuf(&ss,read,0,buf,sizeof(buf));
 substdio_fdbuf(&ssout,write,fd,outbuf,sizeof(outbuf));
 if (substdio_put(&ssout,ufline.s,ufline.len)) goto writeerrs;
 if (substdio_put(&ssout,rpline.s,rpline.len)) goto writeerrs;
 if (substdio_put(&ssout,dtline.s,dtline.len)) goto writeerrs;
 for (;;)
  {
   if (getline2(&ss,&messline,&match,'\n') != 0) 
    { if (flaglocked) seek_trunc(fd,pos); close(fd); temp_read(); }
   if (!match && !messline.len) break;
   if (gfrom(messline.s,messline.len))
     if (substdio_bput(&ssout,">",1)) goto writeerrs;
   if (substdio_bput(&ssout,messline.s,messline.len)) goto writeerrs;
   if (!match)
    {
     if (substdio_bputs(&ssout,"\n")) goto writeerrs;
     break;
    }
  }
 if (substdio_bputs(&ssout,"\n")) goto writeerrs;
 if (substdio_flush(&ssout)) goto writeerrs;
 if (fsync(fd) == -1) goto writeerrs;
 close(fd);
 return;

 writeerrs:
 if (flaglocked) seek_trunc(fd,pos);
 close(fd);
 temp_write();
}

void mailprogram(prog)
char *prog;
{
 int child;
 char *(args[4]);
 int wstat;

 if (seek_begin(0) == -1) temp_rewind();

 switch(child = fork())
  {
   case -1:
     temp_fork();
   case 0:
     args[0] = "sh"; args[1] = "-c"; args[2] = prog; args[3] = 0;
     signal_uninit();
     execvp(*args,args);
     if (errno == error_txtbsy) { err("Text busy. (#4.3.0)\n"); soft(); }
     if (errno == error_nomem) { err("Out of memory. (#4.3.0)\n"); soft(); }
     if (errno == error_io) { err("I/O error. (#4.3.0)\n"); soft(); }
     if (error_temp(errno)) { err("Temporary error. (#4.3.0)\n"); soft(); }
     err("Unable to execute "); err(*args); err(" (#5.2.4)\n");
     hard();
  }

 wait_pid(&wstat,child);
 if (wait_crashed(wstat))
   temp_childcrashed();
 switch(wait_exitcode(wstat))
  {
   case 111: case 120: case 71: case 74: case 75: soft();
   case 0: break;
   case 99: flag99 = 1; break;
   default: hard();
  }
}

unsigned long mailforward_qp = 0;

void mailforward(recips)
char **recips;
{
 struct qqtalk qqt;
 substdio ss;
 int match;

 if (seek_begin(0) == -1) temp_rewind();
 substdio_fdbuf(&ss,read,0,buf,sizeof(buf));

 if (qqtalk_open(&qqt,1) == -1) temp_fork();
 mailforward_qp = qqtalk_qp(&qqt);
 qqtalk_put(&qqt,dtline.s,dtline.len);
 do
  {
   if (getline2(&ss,&messline,&match,'\n') != 0) { qqtalk_fail(&qqt); break; }
   qqtalk_put(&qqt,messline.s,messline.len);
  }
 while (match);
 qqtalk_from(&qqt,ueo.s);
 while (*recips) qqtalk_to(&qqt,*recips++);
 switch(qqtalk_close(&qqt))
  {
   case QQT_TOOLONG: bounce_childperm();
   case QQT_READ: temp_read();
   case 0: return;
   default: temp_child();
  }
}

void bouncexf()
{
 int match;
 substdio ss;

 if (seek_begin(0) == -1) temp_rewind();
 substdio_fdbuf(&ss,read,0,buf,sizeof(buf));
 for (;;)
  {
   if (getline2(&ss,&messline,&match,'\n') != 0) temp_read();
   if (!match) break;
   if (messline.len <= 1)
     break;
   if (messline.len == dtline.len)
     if (!str_diffn(messline.s,dtline.s,dtline.len))
       bounce_loop();
  }
}

void checkhome()
{
 struct stat st;

 if (stat(".",&st) == -1) temp_homestat();
 if (st.st_mode & ALIAS_PATERNALISM) temp_homewritable();
 if (st.st_mode & 01000)
   if (flagdoit) temp_homesticky(); else warn_homesticky();
}

int qmeox(dashowner)
char *dashowner;
{
 struct stat st;

 if (!stralloc_copys(&qme,".qmail")) temp_nomem();
 if (!stralloc_cat(&qme,&dashext)) temp_nomem();
 if (!stralloc_cats(&qme,dashowner)) temp_nomem();
 if (!stralloc_0(&qme)) temp_nomem();
 if (stat(qme.s,&st) == -1)
  {
   if (error_temp(errno)) temp_nfsqmail();
   return -1;
  }
 return 0;
}

int qmeopen(cutable)
int *cutable;
{
 int fd;
 struct stat st;
 int i;

 i = dashext.len;
 for (;;)
  {
   if (!stralloc_copys(&qme,".qmail")) temp_nomem();
   if (!stralloc_catb(&qme,dashext.s,i)) temp_nomem();
   if (i < dashext.len) if (!stralloc_cats(&qme,"-default")) temp_nomem();
   if (!stralloc_0(&qme)) temp_nomem();
   fd = open_read(qme.s);
   if (fd == -1)
    {
     if (error_temp(errno)) temp_nfsqmail();
     if (errno == error_perm) temp_denyqmail();
     if (errno == error_acces) temp_denyqmail();
    }
   else
    {
     if (fstat(fd,&st) == -1) temp_nfsqmail();
     if ((st.st_mode & S_IFMT) == S_IFREG)
      {
       if (st.st_mode & ALIAS_PATERNALISM) temp_qmwritable();
       *cutable = !!(st.st_mode & 0100);
       return fd;
      }
     close(fd);
    }
   if (!i) return -1;
   do
     if (dashext.s[--i] == '-') break;
   while (i);
  }
}

unsigned long count_file = 0;
unsigned long count_forward = 0;
unsigned long count_program = 0;
char count_buf[FMT_ULONG];

char countssbuf[128];
static struct substdio countss = SUBSTDIO_FDBUF(write,1,countssbuf,128);

void count_print()
{
 substdio_puts(&countss,"did ");
 substdio_put(&countss,count_buf,fmt_ulong(count_buf,count_file));
 substdio_puts(&countss,"+");
 substdio_put(&countss,count_buf,fmt_ulong(count_buf,count_forward));
 substdio_puts(&countss,"+");
 substdio_put(&countss,count_buf,fmt_ulong(count_buf,count_program));
 substdio_puts(&countss,"\n");
 if (mailforward_qp)
  {
   substdio_puts(&countss,"qp ");
   substdio_put(&countss,count_buf,fmt_ulong(count_buf,mailforward_qp));
   substdio_puts(&countss,"\n");
  }
 substdio_flush(&countss);
}

void sayit(type,cmd,len)
char *type;
char *cmd;
int len;
{
 substdio_puts(&countss,type);
 substdio_put(&countss,cmd,len);
 substdio_putsflush(&countss,"\n");
}

void main(argc,argv)
int argc;
char **argv;
{
 int opt;
 int i;
 int j;
 int k;
 int fd;
 int numforward;
 char **recips;
 datetime_sec starttime;
 int flagforwardonly;
 char *extx;

 umask(077);
 signal_init();

 if (!env_init()) temp_nomem();

 flagdoit = 1;
 while ((opt = getopt(argc,argv,"nN")) != opteof)
   switch(opt)
    {
     case 'n': flagdoit = 0; break;
     case 'N': flagdoit = 1; break;
     case '?':
     default:
       hard();
    }
 argc -= optind;
 argv += optind;

 if (!(user = *argv++)) usage();
 if (!(homedir = *argv++)) usage();
 if (!(local = *argv++)) usage();
 if (!(dash = *argv++)) usage();
 if (!(ext = *argv++)) usage();
 if (!(host = *argv++)) usage();
 if (!(sender = *argv++)) usage();
 if (*argv) usage();

 if (homedir[0] != '/') usage();
 if (chdir(homedir) == -1) temp_chdir();
 checkhome();

 if (!env_put2("HOST",host)) temp_nomem();
 if (!env_put2("HOME",homedir)) temp_nomem();
 if (!env_put2("USER",user)) temp_nomem();
 if (!env_put2("LOCAL",local)) temp_nomem();

 if (!stralloc_copys(&envrecip,local)) temp_nomem();
 if (!stralloc_cats(&envrecip,"@")) temp_nomem();
 if (!stralloc_cats(&envrecip,host)) temp_nomem();

 if (!stralloc_copy(&foo,&envrecip)) temp_nomem();
 if (!stralloc_0(&foo)) temp_nomem();
 if (!env_put2("RECIPIENT",foo.s)) temp_nomem();

 if (!stralloc_copys(&dtline,"Delivered-To: ")) temp_nomem();
 if (!stralloc_cat(&dtline,&envrecip)) temp_nomem();
 for (i = 0;i < dtline.len;++i) if (dtline.s[i] == '\n') dtline.s[i] = '_';
 if (!stralloc_cats(&dtline,"\n")) temp_nomem();

 if (!stralloc_copy(&foo,&dtline)) temp_nomem();
 if (!stralloc_0(&foo)) temp_nomem();
 if (!env_put2("DTLINE",foo.s)) temp_nomem();

 if (flagdoit)
   bouncexf();

 if (!env_put2("SENDER",sender)) temp_nomem();

 if (!quote2(&foo,sender)) temp_nomem();
 if (!stralloc_copys(&rpline,"Return-Path: <")) temp_nomem();
 if (!stralloc_cat(&rpline,&foo)) temp_nomem();
 for (i = 0;i < rpline.len;++i) if (rpline.s[i] == '\n') rpline.s[i] = '_';
 if (!stralloc_cats(&rpline,">\n")) temp_nomem();

 if (!stralloc_copy(&foo,&rpline)) temp_nomem();
 if (!stralloc_0(&foo)) temp_nomem();
 if (!env_put2("RPLINE",foo.s)) temp_nomem();

 if (!stralloc_copys(&ufline,"From ")) temp_nomem();
 if (*sender)
  {
   int len; int i; char ch;

   len = str_len(sender);
   if (!stralloc_readyplus(&ufline,len)) temp_nomem();
   for (i = 0;i < len;++i)
    {
     ch = sender[i];
     if ((ch == ' ') || (ch == '\t') || (ch == '\n')) ch = '-';
     ufline.s[ufline.len + i] = ch;
    }
   ufline.len += len;
  }
 else
   if (!stralloc_cats(&ufline,"MAILER-DAEMON")) temp_nomem();
 if (!stralloc_cats(&ufline," ")) temp_nomem();
 starttime = now();
 if (!stralloc_cats(&ufline,myctime(starttime))) temp_nomem();

 if (!stralloc_copy(&foo,&ufline)) temp_nomem();
 if (!stralloc_0(&foo)) temp_nomem();
 if (!env_put2("UFLINE",foo.s)) temp_nomem();

 if (!stralloc_copys(&dashext,dash)) temp_nomem();
 if (!stralloc_cats(&dashext,ext)) temp_nomem();
 for (i = 0;i < dashext.len;++i)
   if (dashext.s[i] == '.')
     dashext.s[i] = ':';
 case_lowerb(dashext.s,dashext.len);

 extx = ext;
 if (!env_put2("EXT",extx)) temp_nomem();
 extx += str_chr(extx,'-'); if (*extx) ++extx;
 if (!env_put2("EXT2",extx)) temp_nomem();
 extx += str_chr(extx,'-'); if (*extx) ++extx;
 if (!env_put2("EXT3",extx)) temp_nomem();
 extx += str_chr(extx,'-'); if (*extx) ++extx;
 if (!env_put2("EXT4",extx)) temp_nomem();

 flagforwardonly = 0;
 fd = qmeopen(&flagforwardonly);
 if (fd == -1) if (*dash) bounce_ext();

 if (!stralloc_copys(&ueo,sender)) temp_nomem();
 if (str_diff(sender,""))
   if (str_diff(sender,"#@[]"))
     if (qmeox("-owner") == 0)
      {
       if (qmeox("-owner-default") == 0)
	{
         if (!stralloc_copys(&ueo,local)) temp_nomem();
         if (!stralloc_cats(&ueo,"-owner-@")) temp_nomem();
         if (!stralloc_cats(&ueo,host)) temp_nomem();
         if (!stralloc_cats(&ueo,"-@[]")) temp_nomem();
	}
       else
	{
         if (!stralloc_copys(&ueo,local)) temp_nomem();
         if (!stralloc_cats(&ueo,"-owner@")) temp_nomem();
         if (!stralloc_cats(&ueo,host)) temp_nomem();
	}
      }
 if (!stralloc_0(&ueo)) temp_nomem();
 if (!env_put2("NEWSENDER",ueo.s)) temp_nomem();

 if (!stralloc_ready(&cmds,0)) temp_nomem();
 cmds.len = 0;
 if (fd != -1)
   if (slurpclose(fd,&cmds,256) == -1) temp_nomem();

 if (!cmds.len)
  {
   if (!stralloc_copys(&cmds,ALIAS_EMPTY)) temp_nomem();
   flagforwardonly = 0;
  }
 if (cmds.s[cmds.len - 1] != '\n')
   if (!stralloc_cats(&cmds,"\n")) temp_nomem();

 numforward = 0;
 i = 0;
 for (j = 0;j < cmds.len;++j)
   if (cmds.s[j] == '\n')
    {
     switch(cmds.s[i]) { case '#': case '.': case '/': case '|': break;
       default: ++numforward; }
     i = j + 1;
    }

 recips = (char **) alloc((numforward + 1) * sizeof(char *));
 if (!recips) temp_nomem();
 numforward = 0;

 flag99 = 0;

 i = 0;
 for (j = 0;j < cmds.len;++j)
   if (cmds.s[j] == '\n')
    {
     cmds.s[j] = 0;
     k = j;
     while ((k > i) && (cmds.s[k - 1] == ' ') || (cmds.s[k - 1] == '\t'))
       cmds.s[--k] = 0;
     switch(cmds.s[i])
      {
       case 0: /* k == i */
	 if (i) break;
	 temp_blankline();
       case '#':
         break;
       case '.':
       case '/':
	 ++count_file;
	 if (flagforwardonly) temp_fofile();
	 if (cmds.s[k - 1] == '/')
           if (flagdoit) maildir(cmds.s + i);
           else sayit("maildir ",cmds.s + i,k - i);
	 else
           if (flagdoit) mailfile(cmds.s + i);
           else sayit("mbox ",cmds.s + i,k - i);
         break;
       case '|':
	 ++count_program;
	 if (flagforwardonly) temp_foprog();
         if (flagdoit) mailprogram(cmds.s + i + 1);
         else sayit("program ",cmds.s + i + 1,k - i - 1);
         break;
       case '+':
	 if (!str_diff(cmds.s + i + 1,"list"))
	   flagforwardonly = 1;
	 break;
       case '&':
         ++i;
       default:
	 ++count_forward;
         if (flagdoit) recips[numforward++] = cmds.s + i;
         else sayit("forward ",cmds.s + i,k - i);
         break;
      }
     i = j + 1;
     if (flag99) break;
    }

 if (numforward) if (flagdoit)
  {
   recips[numforward] = 0;
   mailforward(recips);
  }

 count_print();
 _exit(0);
}
